Engine/Component/ComponentButton.cpp

```diff
@@ -1,34 +1,44 @@
 #include "ComponentButton.h"
+#include "ComponentText.h"
 
 #include "Main/Application.h"
 #include "Main/Gameobject.h"
 
-#include "Component/ComponentText.h"
-
 #include "Module/ModuleScene.h"
 
 ComponentButton::ComponentButton() : ComponentUI(ComponentUI::UIType::BUTTON)
 {
-	InitData();
 }
 
 ComponentButton::ComponentButton(GameObject * owner) : ComponentUI(owner, ComponentUI::UIType::BUTTON)
 {
-	InitData();
+	if (owner->transform_2d.is_new)
+	{
+		owner->transform_2d.SetSize(170, 23);
+		owner->transform_2d.SetPosition(&float3(0, 0, 1));
+	}
 }
 
 ComponentButton::~ComponentButton()
 {
 	ComponentUI::~ComponentUI();
 }
 
-void ComponentButton::InitData()
-{
-	owner->transform_2d.SetSize(170, 23);
-}
-
 void ComponentButton::Render(float4x4* projection)
 {
+	owner->transform_2d.position.z = 1;
+	
+	ComponentUI* text = nullptr;
+	for (auto child : owner->children)
+	{
+		text = static_cast<ComponentUI*>(child->GetComponentUI(ComponentUI::UIType::TEXT));
+		if (text)
+		{
+			ComponentTransform2D* text_transform = &text->owner->transform_2d;
+			owner->transform_2d.SetSize(text_transform->rect.Width(), text_transform->rect.Height());
+			break;
+		}
+	}
 	ComponentUI::Render(projection);
 }
```

Comment:
This method and all UI render methods are using the antipatern [call super](https://en.wikipedia.org/wiki/Call_super). This is not desired, because every UI render method is dependant of callin `ComponentUI` `Render` method, and we have to remember to call it every time that we create a new Component for UI. The solution for this would be applying the [template pattern](https://en.wikipedia.org/wiki/Template_method_pattern).